src/Microsoft.Health.Dicom.Core/Features/Query/QueryService.cs

```diff
@@ -58,60 +62,139 @@ public class QueryService : IQueryService
 
         var partitionKey = _contextAccessor.RequestContext.GetPartitionKey();
 
+        Stopwatch stopwatch = new Stopwatch();
+
+        stopwatch.Start();
         QueryResult queryResult = await _queryStore.QueryAsync(partitionKey, queryExpression, cancellationToken);
+        stopwatch.Stop();
+        var filterTime = stopwatch.ElapsedMilliseconds;
 
         if (!queryResult.DicomInstances.Any())
         {
             return new QueryResourceResponse(Array.Empty<DicomDataset>(), queryExpression.ErroneousTags);
         }
 
-        var responseBuilder = new QueryResponseBuilder(queryExpression);
-        IEnumerable<DicomDataset> responseMetadata = null;
+        var responseBuilder = new QueryResponseBuilder(queryExpression, ReturnNewTagDefaults(_contextAccessor.RequestContext.Version));
+
+        stopwatch.Restart();
+        IEnumerable<DicomDataset> instanceMetadata = await GetInstanceMetadataAsync(partitionKey, queryExpression, queryResult, responseBuilder.ReturnTags, cancellationToken);
+        stopwatch.Stop();
+        var resultTime = stopwatch.ElapsedMilliseconds;
+        _logger.LogInformation("QueryService performance filterTimeMilliseconds:{FilterTime}, resultTimeMilliseconds:{ResultTime}", filterTime, resultTime);
+
+        var responseMetadata = instanceMetadata.Select(m => responseBuilder.GenerateResponseDataset(m));
+        return new QueryResourceResponse(responseMetadata, queryExpression.ErroneousTags);
+    }
 
-        if (QueryLimit.ContainsComputedTag(queryExpression.IELevel, responseBuilder.ReturnTags))
+    private static bool ReturnNewTagDefaults(int? apiMajorVersion)
+    {
+        bool useNewDefaults = false;
+        if (apiMajorVersion != null && apiMajorVersion >= 2)
         {
-            var versions = queryResult.DicomInstances.Select(i => i.Version).ToList();
+            useNewDefaults = true;
+        }
+        return useNewDefaults;
+    }
+
+    // Does not handle retrieving the extendedQueryTag indexes right now. Logs are in place to evaluate it in the future.
+    private async Task<IEnumerable<DicomDataset>> GetInstanceMetadataAsync(
+        int partitionKey,
+        QueryExpression queryExpression,
+        QueryResult queryResult,
+        IReadOnlyCollection<DicomTag> returnTags,
+        CancellationToken cancellationToken)
+    {
+        bool getStudyResponse = false, getSeriesResponse = false, getFullMetadata = false;
```

Comment:
It might be worth commenting here that `getFullMetadata` is intended to cover any search for metadata at the Instance level, or any search with `includefield=all` at any level.

Reply:
It also includes anything custom specified in includeField. I will add more comments